src/Core/Auth/Models/Data/GrantTableEntity.cs

```diff
@@ -0,0 +1,65 @@
+﻿using Duende.IdentityServer.Models;
+using Microsoft.Azure.Cosmos.Table;
+
+namespace Bit.Core.Auth.Models.Data;
+
+public class GrantTableEntity : TableEntity, IGrant
+{
+    public GrantTableEntity() { }
+
+    public GrantTableEntity(PersistedGrant pGrant)
+    {
+        var (partitionKey, rowKey) = CreateTableEntityKeys(pGrant.Key);
+        PartitionKey = partitionKey;
+        RowKey = rowKey;
+
+        Key = pGrant.Key;
+        Type = pGrant.Type;
+        SubjectId = pGrant.SubjectId;
+        SessionId = pGrant.SessionId;
+        ClientId = pGrant.ClientId;
+        Description = pGrant.Description;
+        CreationDate = pGrant.CreationTime;
+        ExpirationDate = pGrant.Expiration;
+        ConsumedDate = pGrant.ConsumedTime;
+        Data = pGrant.Data;
+    }
+
+    public GrantTableEntity(IGrant g)
+    {
+        var (partitionKey, rowKey) = CreateTableEntityKeys(g.Key);
+        PartitionKey = partitionKey;
+        RowKey = rowKey;
+
+        Key = g.Key;
+        Type = g.Type;
+        SubjectId = g.SubjectId;
+        SessionId = g.SessionId;
+        ClientId = g.ClientId;
+        Description = g.Description;
+        CreationDate = g.CreationDate;
+        ExpirationDate = g.ExpirationDate;
+        ConsumedDate = g.ConsumedDate;
+        Data = g.Data;
+    }
+
+    public int Id { get; set; }
+    public string Key { get; set; }
+    public string Type { get; set; }
+    public string SubjectId { get; set; }
+    public string SessionId { get; set; }
+    public string ClientId { get; set; }
+    public string Description { get; set; }
+    public DateTime CreationDate { get; set; } = DateTime.UtcNow;
+    public DateTime? ExpirationDate { get; set; }
+    public DateTime? ConsumedDate { get; set; }
+    public string Data { get; set; }
+
+    public static (string partitionKey, string rowKey) CreateTableEntityKeys(string grantKey)
+    {
+        // TODO: How do we want to partition? Or do we? refs:
```

Comment:
I usually look at what we index, but per https://github.com/bitwarden/server/blob/main/src/Sql/Auth/dbo/Tables/Grant.sql there's not much valuable there. The subject ID is technically nullable but never is for us and we put that user ID in there. Separately, would it be valuable to partition on the client ID (which is our type)?

Reply:
⚠️ We want to try a null / empty row key right?

Reply:
yes